design/open/spark-outputcommitter/committer.md

```diff
@@ -89,16 +87,84 @@ overwrite modes: an entire previous table will be deleted on write.
+  object" API.  This adds another link to a staged object, so it is
+  supported as a metadata-only operation.  It occurs once for each
+  successful task, so once for each file partition.
+* **Commit job**:
+  1. Delete all objects in working directories under
+	 `<JOB_BRANCH>/<PATH>/_temporary/`.  All successful task objects were
+	 committed by their tasks and are retained under `<JOB_BRANCH>/<PATH>/`;
+	 all failed task objects are deleted here.  So no temporary objects
+	 remain and all desired objects exist under `<JOB_BRANCH>/<PATH>/`.
+  1. Commit all staged objects.
+  1. Merge the job branch back to the output branch, determining the mode to
+     use by the Hadoop SaveMode.
+
+  This occurs once, but obviously the deletion phase deletes one object for
+  each task attempt.
+
+[^2] We _cannot_ use the job ID to identify the job branch: Spark can create
+  tasks for the same actual write job that use a different job context and
+  ID!
+
+### Requirements from (lakeFS server) Versioning Engine
+
+Assume the number of write partitions $N$ may be several thousands.
+
+1. Support for metadata-only copies.  We want to be able to add another link
+   to an existing object.  We guarantee that that object will be uncommitted
+   and on the same branch.
+
+   As part of GC+, the StageObject API on the lakeFS server is going away,
+   and a new CopyObject API will appear.
+
+   To reduce dependencies, we will use the new API if it is available,
+   otherwise fall back on the old API.  This is safe because lakeFS versions
+   without the upcoming CopyObject can safely StageObject the same object
+   twice.  As an alternative, we would be more than happy to have
+   StageObject limited to support precisely this one use that we need, and
+   that will continue to be supported!
+1. To attain high performance, we will need this inventory of operations to
+   be fast.  Expect each task to write a single object (I have not found any
+   counterexample), so each $\Theta(N)$ is really more like $N$.
+
+   * These operations occur for each write job, so only once or twice:
+     CreateBranch (1), DiffRefs (1, limited to 1 output), DiffBranch (1,
+     limited to 1 output), PutObject (<=1).
+   * These are a small number of batched calls, but in total run in time
+     proportional to the number of partitions, $\Theta(N)$.  Each batch is
+     of course as large as possible (1000): ListObjects, DeleteObjects.
+   * These are single calls that run in time $\Theta(N)$: Commit,
+     MergeIntoBranch.
+   * Each task has to list and copy each result object.  So these are $N$
+     calls to ListObjects that return in total $\Theta(N)$ objects
+     (typically this will be a single object!), and precisely $N$ CopyObject
+     or StageObject calls
+   * Additionally, the Spark writing tasks make lakeFSFS itself perform
+     multiple lakeFS API calls.  Expect $N$ ListObjects with size 1, $\Theta(N)$
+     StatObject, and $\Theta(N)$ PutObject.
```

Comment:
Good point.  #4671 is to add Hadoop-side metrics to lakeFSFS and the OC.  This should help figure out how many calls we actually make, and their performance impact.  And you're right that we may take a performance hit here.

Also I opened #4676 to go look for speedups here, my feeling is that you're right and there are some speedups we can make here by writing more complex code that performs fewer API calls.

Reply:
@arielshaqed If we compare lakeFSFS + FileOutputCommitter with lakeFSFS + LakeFSOutputCommitter:
- Is this correct that the Spark writing tasks make lakeFSFS perform the same lakeFS API calls? 
- Are these calls part of the setup of the tasks? 

What I'm trying to understand is where would the lakeFS OC be different in performance compared to FileOC.

Reply:
LakeFS OC is faster for commits.  Committing a job is _much_ faster because we've already copied everything at the end of each task.  And since it renames correctly (rather than read+write), even that copy is faster.  And LakeFS OC performs a faster deletion of the temporary directory after it commits.

Also when running the recommended v1 algorithm, FOC performs an additional (slow...) copy.